migration/idempotent/017-metric-rollup.sql

```diff
@@ -0,0 +1,277 @@
+                new_rollups_created := new_rollups_created + 1;
+            END IF;
+        END LOOP;
+        RAISE LOG '[Rollup] Created % for rollup name %', new_rollups_created, r.name;
+    END LOOP;
+
+    COMMIT;
+
+    -- Refresh the newly added rollups for entire duration of data.
+    FOR r IN
+        SELECT mr.id as id, rollup.schema_name as schema_name, mt.table_name as table_name
+        FROM _prom_catalog.metric_rollup mr
+                 INNER JOIN _prom_catalog.rollup rollup ON (mr.rollup_id = rollup.id)
+                 INNER JOIN _prom_catalog.metric mt ON (mr.metric_id = mt.id)
+        WHERE mr.refresh_pending = TRUE
+    LOOP
+        CALL public.refresh_continuous_aggregate(r.schema_name || '.' || r.table_name, NULL, NULL);
+        UPDATE _prom_catalog.metric_rollup SET refresh_pending = FALSE WHERE id = r.id;
+        COMMIT;
+    END LOOP;
+END;
+$$
+LANGUAGE PLPGSQL;
+
+-- create_metric_rollup prepares the given rollup. The actual rollup views are created by _prom_catalog.scan_for_new_rollups()
+CREATE OR REPLACE PROCEDURE _prom_catalog.create_rollup(_name TEXT, _resolution INTERVAL, _retention INTERVAL) AS
+$$
+    DECLARE
+        _schema_name constant TEXT := 'ps_' || _name;
+        _exists BOOLEAN;
+
+    BEGIN
+        SELECT EXISTS( SELECT 1 FROM _prom_catalog.rollup WHERE name = _name) INTO _exists;
+        IF _exists THEN
+            RAISE EXCEPTION 'ERROR: cannot create metric rollup for %. REASON: already exists.', name;
+        END IF;
+        -- We do not use IF NOT EXISTS (below) since we want to stop creation process if the schema already exists.
+        -- This forces the user to delete the conflicting schema before proceeding ahead.
+        EXECUTE FORMAT('CREATE SCHEMA %I', _schema_name);
+        INSERT INTO _prom_catalog.rollup(name, schema_name, resolution, retention) VALUES (_name, _schema_name, _resolution, _retention);
+    END;
+$$
+LANGUAGE PLPGSQL;
+GRANT EXECUTE ON PROCEDURE _prom_catalog.create_rollup(text, interval, interval) TO prom_writer;
+
+-- delete_metric_rollup deletes everything related to the given metric rollup label name. It does the following:
+-- 1. Delete the entries in _prom_catalog.rollup
+-- 2. Delete the entries in _prom_catalog.metric_rollup
+-- 3. Delete the actual Cagg views that represent the rollups. This is done by directing deleting the entire rollup schema
+CREATE OR REPLACE PROCEDURE _prom_catalog.delete_rollup(_rollup_name TEXT) AS
+$$
+    DECLARE
+        _rollup_id INTEGER;
+        _rollup_schema_name TEXT;
+
+    BEGIN
+        SELECT id, schema_name INTO _rollup_id, _rollup_schema_name FROM _prom_catalog.rollup WHERE name = _rollup_name;
+        IF _rollup_id IS NULL THEN
+            RAISE EXCEPTION '% rollup not found', _rollup_name;
+        END IF;
+        DELETE FROM _prom_catalog.rollup WHERE id = _rollup_id;
+        DELETE FROM _prom_catalog.metric_rollup WHERE rollup_id = _rollup_id;
```

Comment:
now that there's a foreign key constraint, you need to swap the order in which these are executed. delete from `metric_rollup` first.